### PR TITLE
[feature] #3330: Extend `NumericValue` deserialisation

### LINKED ISSUE #3330: [BUG] Cannot deserialize U128
As per title. Issue spans from LTS to current ([726f5eabf65a79ea618b4fce62a09cee7a5b13d1](https://github.com/hyperledger/iroha/commit/726f5eabf65a79ea618b4fce62a09cee7a5b13d1)).
Easiest way to reproduce: start iroha with modified `genesis.json`.

<details><summary> <code>genesis.json</code> for LTS</summary>
<p>

```javascript
{
  "transactions": [
    {
      "isi": [
        {
          "Register": {
            "Identifiable": {
              "NewDomain": {
                "id": "wonderland",
                "logo": null,
                "metadata": {
                  "key": {
                    "U128": 1
                  }
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "NewAccount": {
                "id": "alice@wonderland",
                "signatories": [
                  "ed01207233bfc89dcbd68c19fde6ce6158225298ec1131b6a130d1aeb454c1ab5183c0"
                ],
                "metadata": {
                  "key": {
                    "U128": 1
                  }
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "NewAccount": {
                "id": "bob@wonderland",
                "signatories": [
                  "ed01207233bfc89dcbd68c19fde6ce6158225298ec1131b6a130d1aeb454c1ab5183c0"
                ],
                "metadata": {
                  "key": {
                    "U128": 1
                  }
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "NewAssetDefinition": {
                "id": "rose#wonderland",
                "value_type": "Quantity",
                "mintable": "Infinitely",
                "metadata": {}
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "NewDomain": {
                "id": "garden_of_live_flowers",
                "logo": null,
                "metadata": {}
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "NewAccount": {
                "id": "carpenter@garden_of_live_flowers",
                "signatories": [
                  "ed01207233bfc89dcbd68c19fde6ce6158225298ec1131b6a130d1aeb454c1ab5183c0"
                ],
                "metadata": {}
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "NewAssetDefinition": {
                "id": "cabbage#garden_of_live_flowers",
                "value_type": "Quantity",
                "mintable": "Infinitely",
                "metadata": {}
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "PermissionTokenDefinition": {
                "id": "can_unregister_asset_with_definition",
                "params": {
                  "asset_definition_id": "Id"
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "PermissionTokenDefinition": {
                "id": "can_burn_asset_with_definition",
                "params": {
                  "asset_definition_id": "Id"
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "PermissionTokenDefinition": {
                "id": "can_burn_user_assets",
                "params": {
                  "asset_id": "Id"
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "PermissionTokenDefinition": {
                "id": "can_set_key_value_in_user_assets",
                "params": {
                  "asset_id": "Id"
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "PermissionTokenDefinition": {
                "id": "can_remove_key_value_in_user_assets",
                "params": {
                  "asset_id": "Id"
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "PermissionTokenDefinition": {
                "id": "can_set_key_value_in_user_metadata",
                "params": {
                  "account_id": "Id"
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "PermissionTokenDefinition": {
                "id": "can_remove_key_value_in_user_metadata",
                "params": {
                  "account_id": "Id"
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "PermissionTokenDefinition": {
                "id": "can_set_key_value_in_asset_definition",
                "params": {
                  "asset_definition_id": "Id"
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "PermissionTokenDefinition": {
                "id": "can_remove_key_value_in_asset_definition",
                "params": {
                  "asset_definition_id": "Id"
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "PermissionTokenDefinition": {
                "id": "can_mint_user_asset_definitions",
                "params": {
                  "asset_definition_id": "Id"
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "PermissionTokenDefinition": {
                "id": "can_transfer_user_assets",
                "params": {
                  "asset_id": "Id"
                }
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "PermissionTokenDefinition": {
                "id": "can_transfer_only_fixed_number_of_times_per_period",
                "params": {
                  "count": "U32",
                  "period": "U128"
                }
              }
            }
          }
        },
        {
          "Mint": {
            "object": {
              "U32": 13
            },
            "destination_id": {
              "Id": {
                "AssetId": "rose##alice@wonderland"
              }
            }
          }
        },
        {
          "Mint": {
            "object": {
              "U32": 44
            },
            "destination_id": {
              "Id": {
                "AssetId": "cabbage#garden_of_live_flowers#alice@wonderland"
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "PermissionTokenDefinition": {
                "id": "allowed_to_do_stuff",
                "params": {}
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "NewRole": {
                "inner": {
                  "id": "USER_METADATA_ACCESS",
                  "permissions": [
                    {
                      "definition_id": "can_remove_key_value_in_user_metadata",
                      "params": {
                        "account_id": {
                          "Id": {
                            "AccountId": "alice@wonderland"
                          }
                        }
                      }
                    },
                    {
                      "definition_id": "can_set_key_value_in_user_metadata",
                      "params": {
                        "account_id": {
                          "Id": {
                            "AccountId": "alice@wonderland"
                          }
                        }
                      }
                    }
                  ]
                }
              }
            }
          }
        },
        {
          "Grant": {
            "object": {
              "PermissionToken": {
                "definition_id": "allowed_to_do_stuff",
                "params": {}
              }
            },
            "destination_id": {
              "Id": {
                "AccountId": "alice@wonderland"
              }
            }
          }
        },
        {
          "Register": {
            "Identifiable": {
              "NewRole": {
                "inner": {
                  "id": "staff_that_does_stuff_in_genesis",
                  "permissions": [
                    {
                      "definition_id": "allowed_to_do_stuff",
                      "params": {}
                    }
                  ]
                }
              }
            }
          }
        }
      ]
    }
  ]
}
``` 

</p>
</details>

<details><summary> <code>genesis.json</code> for DEV</summary>
<p>

[gist link](https://gist.github.com/QuentinI/191c32516b59f2305fcd1989662c4c29)

</p>
</details> 

Both files attached differ only with value type of example metadata, which is U128 instead of String. They were created by modifying relevant line in kagami and re-generating genesis (genesis for dev is also edited by hand to correct for another bug we have on dev)

### LINKED ISSUE #3539: [BUG] Fails to Deserialize Genesis Block with U32 metadata
### OS and Environment

MacOS, Docker Hub

### GIT commit hash

b1855454

### Minimum working example / Steps to reproduce

Replace the default registration instruction for Alice in the genesis block with a new one:
```json
{
    "Register": {
      "NewAccount": {
        "id": "alice@wonderland",
        "signatories": [
          "ed01207233BFC89DCBD68C19FDE6CE6158225298EC1131B6A130D1AEB454C1AB5183C0"
        ],
        "metadata": {
          "key": {
            "String": "value"
          },
          "u128Key": {
            "U32": "4294967294"
          }
        }
      }
    }
},
```
2. `docker compose up`

### Actual result

```
2023-05-29 15:06:10 2023-05-29T11:06:10.381410Z  WARN iroha: The configuration parameter `DISABLE_PANIC_TERMINAL_COLORS` is deprecated. Set `TERMINAL_COLORS=false` instead. 
2023-05-29 15:06:10 2023-05-29T11:06:10.383227Z  INFO iroha: Hyperledgerいろは2にようこそ！(translation) Welcome to Hyperledger Iroha 2.0.0-pre-rc.13! git_commit_sha="b7e80f04c210c52aeff261c1902c6888bea29b97"
2023-05-29 15:06:10 Error: Failed to deserialize raw genesis block from "/config/genesis.json"
2023-05-29 15:06:10 
2023-05-29 15:06:10 Caused by:
2023-05-29 15:06:10     data did not match any variant of untagged enum DeserializeHelper at line 34 column 7
2023-05-29 15:06:10 
2023-05-29 15:06:10 Location:
2023-05-29 15:06:10     /iroha/genesis/src/lib.rs:216:75
```

### Expected result

Iroha should launch without any issues, and the new registration instruction for Alice should be processed correctly.

### Who can help to reproduce?

@astrokov7

### PR DESCRIPTION
**Pull request by appetrosyan**
## Description

Implemented manual deserialisation for `NumericValue` to circumvent https://github.com/serde-rs/json/issues/846. 

All `NumericValue` instances can now be deserialised from a string directly. 

All instances of `NumericValue::U128` are now preferentially serialised as tagged string-valued enums. Literal syntax deserialisation is supported but discouraged, given the slowness of processing. 

### Linked issue

Closes #3330 
Closes #3539 

### Benefits

More flexible syntax

### Drawbacks

Potentially slower deserialsation

### Checklist

- [x] Self-review
- [x] Implementation
- [x] Review comments
- [x] tests
  - [x] ser
  - [x] de
  - [x] ToString
  - [x] FromString
  - [x] tagged

### CODE REVIEW

File: data_model/derive/src/partially_tagged.rs [hunk 1]
```diff
@@ -262,19 +271,34 @@ pub fn impl_partially_tagged_deserialize(enum_: &PartiallyTaggedEnum) -> TokenSt
...
+                            }
+							message+="\n--------------\n";
                             for (error, candidate_type) in errors {
-                                message +=". ";
                                 message +="Candidate `";
                                 message += &candidate_type;
                                 message += "` failed to deserialize with error: ";
-                                message += &error.to_string();
+                                message += &error;
+								message += "\n";
```
Comment (mversic) [inline data_model/derive/src/partially_tagged.rs:301]: lots of lines with tabs in this file

Comment (appetrosyan) [reply data_model/derive/src/partially_tagged.rs]: Fixed.

File: data_model/src/numeric.rs [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+                }
+            }
+            (Discriminants::U128, data) => {
+                let val = data.newtype_variant::<Value>()?;
+                // NOTE: This is slower than deserializing from a
+                // string, because this uses an arbitrary precision
+                // conversion instead of parsing directly.
+                if let Ok(num) = u128::deserialize(&val) {
+                    #[cfg(debug_assertions)]
+                    eprintln!("`u128` literals are slow.");
```
[the remark this answers is not in the document]
Comment (appetrosyan) [reply data_model/src/numeric.rs:190]: Apparently, it might not be initialised in all cases. I'll investigate and find a way to report it to the user.

File: data_model/src/metadata.rs [hunk 3]
```diff
@@ -303,6 +303,7 @@ mod tests {
         assert!(metadata
             .nested_insert_with_limits(&empty_path, "0".to_owned().into(), Limits::new(12, 12))
             .is_err());
+        #[cfg(feature = "transparent_api")]
```
[the remark this answers is not in the document]
Comment (appetrosyan) [reply data_model/src/metadata.rs:306]: If you run `cargo test` from workspace root, the resolver enables `transparent_api` and this is not a compile error. 

However, the function `nested_remove` is not defined without the `transparent_api` and the test fails to compile if you run `cargo test` from the `data_model` dir.

File: data_model/src/numeric.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (appetrosyan) [reply data_model/src/numeric.rs:190]: Actually, after a bit of thought, the warning is not going to do much useful here. I'll remove it entirely.

File: data_model/src/numeric.rs [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+}
+
+impl From<FixedPointOperationError> for ParseNumericError {
+    fn from(value: FixedPointOperationError) -> Self {
+        Self::ParseFixed(value)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
[the remark this answers is not in the document]
Comment (appetrosyan) [reply data_model/src/numeric.rs:263]: Ticket created. The edge case coverage should be fine for now.

File: data_model/src/numeric.rs [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+    // an `untagged` `StringOrNumber` enum.
+    fn visit_enum<E>(self, value: E) -> Result<NumericValue, E::Error>
+    where
+        E: serde::de::EnumAccess<'de>,
+    {
+        use serde_json::Value;
+
+        match value.variant::<Discriminants>()? {
+            (Discriminants::U32, data) => {
+                let val = data.newtype_variant::<Value>()?;
```
Comment (Erigara) [inline data_model/src/numeric.rs:161]: We use `serde_json::Value` because we want to represent `String` and number in single enum?

**Review by mversic — APPROVED** [review]

File: data_model/src/numeric.rs (hunk 5 above)
Comment (appetrosyan) [reply data_model/src/numeric.rs:161]: We use this so that `serde` sees the u128 literal without defining another useeless struct with a visitor. This is an extremely ugly workaround, but it could have been much worse.

Comment (Erigara) [reply data_model/src/numeric.rs:161]: Let's leave it as is for now then

**Review by Erigara — APPROVED** [review]
